src/main/scala/chisel3/util/MemInterface.scala

```diff
@@ -0,0 +1,337 @@
+import chisel3.experimental.SourceInfo
+import chisel3.internal.sourceinfo.{MemTransform, SourceInfoTransform}
+import scala.language.reflectiveCalls
+import scala.language.experimental.macros
+
+/** A bundle of signals representing a read memory port.
+  *
+  * @tparam tpe The data type of the memory port
+  * @param addrWidth The width of the address signal
+  */
+class MemRdPortInterface[T <: Data](tpe: T, addrWidth: Int) extends Bundle {
+  val addr = Input(UInt(addrWidth.W))
+  val enable = Input(Bool())
+  val readValue = Output(tpe)
+}
+
+/** A bundle of signals representing a write memory port.
+  *
+  * @tparam tpe The data type of the memory port
+  * @param addrWidth The width of the address signal
+  * @param withMask Whether this read/write port interface should have an optional mask.
+  *
+  * @note `withMask` is only valid if tpe is a Vec; if this is not the case no mask will be initialized
+  *       regardless of the value of `withMask`.
+  */
+class MemWrPortInterface[T <: Data](tpe: T, addrWidth: Int, withMask: Boolean) extends Bundle {
+  val maskSize = tpe match {
+    case vec: Vec[_] => vec.size
+    case _ => 0
+  }
+
+  val addr = Input(UInt(addrWidth.W))
+  val enable = Input(Bool())
+  val writeValue = Input(tpe)
+  val mask: Option[Vec[Bool]] = if (withMask && maskSize > 0) Some(Input(Vec(maskSize, Bool()))) else None
+}
+
+/** A bundle of signals representing a read/write memory port.
+  *
+  * @tparam tpe The data type of the memory port
+  * @param addrWidth The width of the address signal
+  * @param withMask Whether this read/write port interface should have an optional mask.
+  *
+  * @note `withMask` is only valid if tpe is a Vec; if this is not the case no mask will be initialized
+  *       regardless of the value of `withMask`.
+  */
+class MemRdWrPortInterface[T <: Data](tpe: T, addrWidth: Int, withMask: Boolean) extends Bundle {
+  val maskSize = tpe match {
+    case vec: Vec[_] => vec.size
+    case _ => 0
+  }
+
+  val addr = Input(UInt(addrWidth.W))
+  val enable = Input(Bool())
+  val isWrite = Input(Bool())
+  val readValue = Output(tpe)
+  val writeValue = Input(tpe)
+  val mask: Option[Vec[Bool]] = if (withMask && maskSize > 0) Some(Input(Vec(maskSize, Bool()))) else None
+}
+
+/** A IO bundle of signals connecting to the ports of a wrapped `SyncReadMem`, as requested by
+  * `SyncReadMem.interface`.
+  *
+  * @tparam tpe The data type of the memory port
+  * @param width The width of the address wires of each port
+  * @param numRd The number of read ports
+  * @param numWr The number of write ports
+  * @param numRdWr The number of read/write ports
+  */
+class MemInterface[T <: Data](tpe: T, addrWidth: Int, numRd: Int, numWr: Int, numRdWr: Int, withMask: Boolean)
+    extends Bundle {
+  if (withMask) {
+    require(
+      tpe.isInstanceOf[Vec[_]],
+      s"masked writes require that MemInterface is instantiated with a data type of Vec (got $tpe instead)"
+    )
+  }
```

Comment:
I see you have the check here rather than in each port. That is fine, but if we want to do that, we need to make the other constructors package private so that you can only construct them through this type safety.